### PR TITLE
Give up when ExchangeFinder exhausts all routes.

### PR DESCRIPTION
**Pull request by dave-r12**
https://github.com/square/okhttp/issues/3308

### CODE REVIEW

**Comment by dave-r12** [comment]
There is a (extremely rare?) scenario where we may make a connection to a server that is then flagged as unhealthy. Currently 4.4.0 and master will throw a `NoSuchElementException`.

I ran this test against 4.3.1 and we appear to go an infinite loop. I'm not sure the right action to take in this scenario. Let me know your thoughts.

**Comment by yschimke** [comment]
Not sure about the right fix.  Seems like worth fixing in 4.5, probably commit this behind @Ignore first?

**Comment by swankjesse** [comment]
Great test

**Comment by yschimke** [comment]
This PR is the hit sequel to

image

File: okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt [hunk 1]
```diff
@@ -113,13 +113,23 @@ class ExchangeFinder(
...
+        return candidate
       }
 
-      return candidate
+      // If it isn't, take it out of the pool.
+      candidate.noNewExchanges()
+
+      // Make sure we have some routes left to try. One example where we may exhaust all the routes
+      // would happen if we made a new connection and it immediately is detected as unhealthy.
+      synchronized(connectionPool) {
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt]: Should we put a 
    connectionPool.assertThreadDoesntHoldLock()

At the top of this method?  I find it really really useful to document the current expected state of locks whenever we then open into a lock.

File: okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt [hunk 2]
```diff
@@ -113,13 +113,23 @@ class ExchangeFinder(
...
+        val routesLeft = routeSelection?.hasNext() ?: true
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt]: It seems like line 108 must have made routeSelection non null, so can we change to !!

File: okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt [hunk 3]
```diff
@@ -113,13 +113,23 @@ class ExchangeFinder(
...
+        val routesSelectionLeft = routeSelector?.hasNext() ?: true
+        if (nextRouteToTry == null && !routesLeft && !routesSelectionLeft) {
+          throw IOException("exhausted all routes")
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt]: We can probably be more expressive in this error, it's an interesting case when routes are immediately unhealthy.

**Comment by yschimke** [comment]
Nice PR!

File: okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt [hunk 4]
```diff
@@ -113,13 +113,23 @@ class ExchangeFinder(
...
-      return candidate
+      // If it isn't, take it out of the pool.
+      candidate.noNewExchanges()
+
+      // Make sure we have some routes left to try. One example where we may exhaust all the routes
+      // would happen if we made a new connection and it immediately is detected as unhealthy.
+      synchronized(connectionPool) {
+        val routesLeft = routeSelection?.hasNext() ?: true
+        val routesSelectionLeft = routeSelector?.hasNext() ?: true
+        if (nextRouteToTry == null && !routesLeft && !routesSelectionLeft) {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt]: I’d be tempted to change this to short circuit more aggressively?

```
      // Make sure we have some routes left to try. One example where we may exhaust all the routes
      // would happen if we made a new connection and it immediately is detected as unhealthy.
      synchronized(connectionPool) {
        if (nextRouteToTry != null) continue

        val routesLeft = routeSelection?.hasNext() ?: true
        if (routesLeft) continue

        val routesSelectionLeft = routeSelector?.hasNext() ?: true
        if (routesSelectionLeft) continue

        throw IOException("exhausted all routes")
      }
```

**Comment by swankjesse** [comment]
I’m gonna merge this as-is because it fixes a big bug, and is already great.

Please address feedback in follow-up PRs as necessary!

File: okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt (hunk 2 above)
Comment (dave-r12) [reply okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt]: These will be `null` if the call had a previous connection we could use. Here's a test that demonstrates it:

```java
    server.enqueue(new MockResponse().setResponseCode(401));
    server.enqueue(new MockResponse().setResponseCode(200));
    AtomicReference<Connection> ref = new AtomicReference<>();
    EventListener listener = new EventListener() {
      @Override public void connectionAcquired(Call call, Connection connection) {
        ref.set(connection);
      }
    };

    client = client.newBuilder()
        .eventListener(listener)
        .authenticator((route, response) -> {
          Util.closeQuietly(ref.get().socket());
          return response.request();
        })
        .build();

    executeSynchronously("/").assertCode(200);
```

**Comment by Vinod217** [comment]
Hi @swankjesse & others, We are also facing this issue with okhttp3 3.12.8 version, Could you please confirm if this issue fixed in any latest okhttp3 version?

okhttp3- 3.12.8
Retrofit- 2.6.1

Logs:-
Fatal Exception: java.lang.NoSuchMethodError
No virtual method ,??Վ�蕚õ0ޅ�V��BsK촜ϯ��+Г�?3??_S?#*烔H奼�cMݗ�Ӷɞͷtey�Nԣݛ?java/net/URI;)Ljava/util/List; in class Ljava/net/ProxySelector; or its super classes (declaration of 'java.net.ProxySelector' appears in /system/framework/core-oj.jar)
okhttp3.internal.connection.RouteSelector.resetNextProxy (RouteSelector.java:129)
okhttp3.internal.connection.RouteSelector. (RouteSelector.java:63)
okhttp3.internal.connection.StreamAllocation. (StreamAllocation.java:101)
okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept (RetryAndFollowUpInterceptor.java:113)
okhttp3.internal.http.RealInterceptorChain.proceed (RealInterceptorChain.java:147)
okhttp3.internal.http.RealInterceptorChain.proceed (RealInterceptorChain.java:121)
okhttp3.RealCall.getResponseWithInterceptorChain (RealCall.java:257)

Thanks in advance.

**Pull request merged**